lib/puppet/configurer.rb

```diff
@@ -417,23 +411,17 @@ def run_internal(options)
   private :run_internal
 
   def find_functional_server
-    Puppet.settings[:server_list].each do |server|
-      host = server[0]
-      port = server[1] || Puppet[:masterport]
-      begin
-        ssl_context = Puppet.lookup(:ssl_context)
-        http = Puppet::Network::HttpPool.connection(host, port.to_i, ssl_context: ssl_context)
-        response = http.get('/status/v1/simple/master')
-        return [host, port] if response.is_a?(Net::HTTPOK)
-
-        Puppet.debug(_("Puppet server %{host}:%{port} is unavailable: %{code} %{reason}") %
-                     { host: host, port: port, code: response.code, reason: response.message })
-      rescue => detail
-        #TRANSLATORS 'server_list' is the name of a setting and should not be translated
-        Puppet.debug _("Unable to connect to server from server_list setting: %{detail}") % {detail: detail}
-      end
+    begin
+      session = Puppet.lookup(:http_session)
+      service = session.route_to(:puppet)
+      [service.url.host, service.url.port]
+    rescue Puppet::HTTP::ResponseError => e
+      Puppet.debug(_("Puppet server %{host}:%{port} is unavailable: %{code} %{reason}") %
+                   { host: e.response.url.host, port: e.response.url.port, code: e.response.code, reason: e.response.reason })
+    rescue => detail
+      #TRANSLATORS 'server_list' is the name of a setting and should not be translated
+      Puppet.debug _("Unable to connect to server from server_list setting: %{detail}") % {detail: detail}
     end
-    [nil, nil]
```

Comment:
The code works without this line, but I think it's more explicit what's happening with this line present. Let me add it back.

Reply:
Updated